README.md
pyHIBP (pyHave I Been Pwned)
============================
[![image](https://img.shields.io/pypi/v/pyHIBP.svg)](https://pypi.org/project/pyHIBP/)
[![image](https://img.shields.io/pypi/l/pyHIBP.svg)](https://pypi.org/project/pyHIBP/)
[![image](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gl/kitsunix%2FpyHIBP%2FpyHIBP-binder/master)


A Python interface to Troy Hunt's 'Have I Been Pwned?' (HIBP) public API. A full reference to the API
specification can be found at the [HIBP API Reference](https://haveibeenpwned.com/API/v2).

This module detects when the rate limit of the API has been hit, and raises a RuntimeError when the limit is exceeded,
or when another API-defined error condition is encountered based on the submitted data. Calls
to the module returning Boolean `True` or the object as decoded from the API query (currently, lists), represent
a detection that a breached account/paste/password was found; Boolean `False` means that the item was not found.

Note that the `pwnedpasswords` API backend does not have a rate limit. If you are intending to bulk-query passwords or
hashes, you may also consider downloading the raw data files accessible via the [Pwned Passwords](https://haveibeenpwned.com/Passwords) page.

Installing
----------
```bash
$ pip install pyhibp
```

Example usage
-------------
For an interactive example, check out the Jupyter Notebook for [`pyhibp`](https://mybinder.org/v2/gl/kitsunix%2FpyHIBP%2FpyHIBP-binder/master?filepath=/pyHIBP.ipynb),
as well as [`pyhibp.pwnedpasswords`](https://mybinder.org/v2/gl/kitsunix%2FpyHIBP%2FpyHIBP-binder/master?filepath=/pyHIBP.pwnedpasswords.ipynb).

```python
import pyhibp
from pyhibp import pwnedpasswords as pw

# Check a password to see if it has been disclosed in a public breach corpus
resp = pw.is_password_breached(password="secret")
if resp:
    print("Password breached!")
    print("This password was used {0} time(s) before.".format(resp))

# Get breaches that affect a given account
resp = pyhibp.get_account_breaches(account="test@example.com", truncate_response=True)

# Get all breach information
resp = pyhibp.get_all_breaches()

# Get a single breach
resp = pyhibp.get_single_breach(breach_name="Adobe")

# Get pastes affecting a given email address
resp = pyhibp.get_pastes(email_address="test@example.com")

# Get data classes in the HIBP system
resp = pyhibp.get_data_classes()
```

Developing
----------
This project is currently intended to be compatible with Python 2 and Python 3. As such, we use virtual environments via `pipenv`.
To develop or test, execute the following:

```bash
# Install the prerequisite virtual environment provider
$ pip install pipenv
# Initialize the pipenv environment and install the module within it
$ make dev
# To run PEP8, tests, and check the manifest
$ make tox
```

Other commands can be found in the `Makefile`.

Goals
-----
- Synchronize to the latest HIBP API(s), implementing endpoint accessing functions where it makes sense. For instance,
  in the interest of security, the ability to submit a SHA-1 to the Pwned Passwords endpoint is not implemented. See
  "Regarding password checking" below for further details.
- For breaches and pastes, act as an intermediary; return the JSON as received from the service.

Regarding password checking
---------------------------
- For passwords, the option to supply a plaintext password to check is provided as an implementation convenience.
- For added security, `pwnedpasswords.is_password_breached()` only transmits the first five characters of the SHA-1
  hash to the Pwned Passwords API endpoint; a secure password will remain secure without disclosing the full hash.


CHANGELOG
v2.0.1 (2018-MM-DD)
-------------------
- **Deprecation warning**: The function `is_password_breached` has moved from `pyHIBP` to `pwnedpasswords`. A wrapper has
been left to ease transition to the new function.
- The Pwned Passwords API has been released, and as such the following new functions have been added...
- `is_password_breached` in `pwnedpasswords` now returns Integer zero (0) if a password was not breached, and an Integer
count of the number of times the password was found in the Pwned Password corpus if the password's hash was found.
- `range_search` supplies only five characters of the SHA-1 hash of a password to the Pwned Passwords server, permitting
a potentially unbreached password to remain just that: unbreached.

v2.0.0 (2018-02-01)
-------------------
- Initial release.


test/conftest.py
import time

import pytest


@pytest.fixture(autouse=True)
def rate_limit():
    # The HIBP API has a ratelimit of 1500ms. Sleep for 2 seconds.
    time.sleep(2)


test/test_pyHIBP.py
import hashlib
import time

import pytest

import pyHIBP


TEST_ACCOUNT = "test@example.com"
TEST_DOMAIN = "adobe.com"
TEST_DOMAIN_NAME = "Adobe"
TEST_PASSWORD = "password"
TEST_PASSWORD_SHA1_HASH = hashlib.sha1(TEST_PASSWORD.encode('utf-8')).hexdigest()
# At least, I doubt someone would have used this (only directly specifying here for deterministic tests...)
TEST_PASSWORD_LIKELY_NOT_COMPROMISED = "&Q?t{%i|n+&qpyP/`/Lyr3<rK|N/M//;u^!fnR+j'lM)A+IGcgRGs[6mLY7yV-|x0bYB&L.JyaJ"
TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH = hashlib.sha1(TEST_PASSWORD_LIKELY_NOT_COMPROMISED.encode('utf-8')).hexdigest()


@pytest.fixture(autouse=True)
def rate_limit():
    # The HIBP API has a ratelimit of 1500ms. Sleep for 2 seconds.
    time.sleep(2)


class TestGetBreaches(object):
    def test_get_breaches_account(self):
        # get_account_breaches(account=TEST_ACCOUNT, domain=None, truncate_response=False, include_unverified=False):
        resp = pyHIBP.get_account_breaches(account=TEST_ACCOUNT)
        assert isinstance(resp, list)
        # As of a manual test, there were 46 accounts for the test@example.com; so >=20 is safe.
        assert len(resp) >= 20
        assert isinstance(resp[0], dict)

    def test_get_breaches_account_with_domain(self):
        # get_account_breaches(account=TEST_ACCOUNT, domain=TEST_DOMAIN, truncate_response=False, include_unverified=False):
        resp = pyHIBP.get_account_breaches(account=TEST_ACCOUNT, domain=TEST_DOMAIN)
        assert isinstance(resp, list)
        # We're limiting the domain; so we only expect one breach to be returned
        assert len(resp) == 1
        assert isinstance(resp[0], dict)
        assert resp[0]['Name'] == TEST_DOMAIN_NAME

    def test_get_breaches_account_with_truncation(self):
        # get_account_breaches(account=TEST_ACCOUNT, domain=None, truncate_response=True, include_unverified=False):
        resp = pyHIBP.get_account_breaches(account=TEST_ACCOUNT, truncate_response=True)
        assert isinstance(resp, list)
        assert len(resp) >= 20
        assert isinstance(resp[0], dict)
        # The individual dicts are only the name of the breached website (since we're truncating)
        item = resp[0]
        assert len(item) == 1
        assert 'Name' in item
        assert 'DataClasses' not in item

    def test_get_breaches_retrieve_all_breaches_with_unverified(self):
        # get_account_breaches(account=TEST_ACCOUNT, domain=None, truncate_response=False, include_unverified=True):
        resp = pyHIBP.get_account_breaches(account=TEST_ACCOUNT, include_unverified=True)
        assert isinstance(resp, list)
        assert len(resp) > 50
        has_unverified = False
        for item in resp:
            if not item['IsVerified']:
                has_unverified = True
                # If we see any unverified items, that's enough.
                break
        assert has_unverified

    def test_get_breaches_return_false_if_no_accounts(self):
        # get_account_breaches(account=TEST_PASSWORD_SHA1_HASH, domain=None, truncate_response=False, include_unverified=False):
        resp = pyHIBP.get_account_breaches(account=TEST_PASSWORD_SHA1_HASH)
        assert not resp
        assert isinstance(resp, bool)

    def test_get_breaches_raise_if_account_is_not_specified(self):
        # get_account_breaches(account=1, domain=None, truncate_response=False, include_unverified=False):
        with pytest.raises(AttributeError) as excinfo:
            # Will raise because the account must be a string (specifically, six.text_type)
            pyHIBP.get_account_breaches(account=None)
        assert "The account parameter must be specified, and must be a string" in str(excinfo.value)

    def test_get_breaches_raise_if_account_is_not_string(self):
        # get_account_breaches(account=1, domain=None, truncate_response=False, include_unverified=False):
        with pytest.raises(AttributeError) as excinfo:
            # Will raise because the account must be a string (specifically, six.text_type)
            pyHIBP.get_account_breaches(account=1)
        assert "The account parameter must be specified, and must be a string" in str(excinfo.value)

    def test_get_breaches_raise_if_domain_is_not_string(self):
        # get_account_breaches(account=TEST_ACCOUNT, domain=1, truncate_response=False, include_unverified=False):
        with pytest.raises(AttributeError) as excinfo:
            # Will raise because the domain must be a string (specifically, six.text_type)
            pyHIBP.get_account_breaches(account=TEST_ACCOUNT, domain=1)
        assert "The domain parameter, if specified, must be a string" in str(excinfo.value)


class TestGetAllBreaches(object):
    def test_get_all_breaches(self):
        # def get_all_breaches(domain=None):
        resp = pyHIBP.get_all_breaches()
        assert isinstance(resp, list)
        assert len(resp) > 50
        assert isinstance(resp[0], dict)

    def test_get_all_breaches_filter_to_domain(self):
        # def get_all_breaches(domain=TEST_DOMAIN):
        resp = pyHIBP.get_all_breaches(domain=TEST_DOMAIN)
        assert isinstance(resp, list)
        # There can be multiple breaches in the system for a given domain
        assert len(resp) >= 1
        assert isinstance(resp[0], dict)
        assert resp[0]['Name'] == TEST_DOMAIN_NAME

    def test_get_all_breaches_false_if_domain_does_not_exist(self):
        resp = pyHIBP.get_all_breaches(domain=TEST_PASSWORD_SHA1_HASH)
        assert not resp
        assert isinstance(resp, bool)

    def test_get_all_breaches_raise_if_not_string(self):
        # def get_all_breaches(domain=1):
        with pytest.raises(AttributeError) as excinfo:
            # Will raise because the domain must be a string (specifically, six.text_type)
            pyHIBP.get_all_breaches(domain=1)
        assert "The domain parameter, if specified, must be a string" in str(excinfo.value)


class TestGetSingleBreach(object):
    def test_get_single_breach(self):
        # pyHIBP.get_single_breach(breach_name=TEST_DOMAIN_NAME)
        resp = pyHIBP.get_single_breach(breach_name=TEST_DOMAIN_NAME)
        assert isinstance(resp, dict)
        assert resp['Name'] == TEST_DOMAIN_NAME

    def test_get_single_breach_when_breach_does_not_exist(self):
        # pyHIBP.get_single_breach(breach_name="ThisShouldNotExist")
        resp = pyHIBP.get_single_breach(breach_name="ThisShouldNotExist")
        # Boolean False will be returned from the above (as there is no breach named what we gave it).
        assert not resp

    def test_get_single_breach_raise_when_breach_name_not_specified(self):
        # pyHIBP.get_single_breach()
        with pytest.raises(AttributeError) as excinfo:
            # Will error because the breach_name must be specified
            pyHIBP.get_single_breach()
        assert "The breach_name must be specified, and be a string" in str(excinfo.value)

    def test_get_single_breach_raise_when_breach_name_is_not_a_string(self):
        # pyHIBP.get_single_breach(breach_name=1)
        with pytest.raises(AttributeError) as excinfo:
            # Will raise because the breach_name must be a string (specifically, six.text_type)
            pyHIBP.get_single_breach(breach_name=1)
        assert "The breach_name must be specified, and be a string" in str(excinfo.value)


class TestGetPastes(object):
    def test_get_pastes(self):
        # pyHIBP.get_pastes(email_address=TEST_ACCOUNT):
        resp = pyHIBP.get_pastes(email_address=TEST_ACCOUNT)
        # The return value is a list, containing multiple dicts (1 or more)
        assert isinstance(resp, list)
        for item in resp:
            assert isinstance(item, dict)

    def test_get_pastes_return_false_if_no_account(self):
        # pyHIBP.get_pastes(email_address=TEST_ACCOUNT):
        resp = pyHIBP.get_pastes(email_address=TEST_PASSWORD_SHA1_HASH + "@example.invalid")
        assert not resp
        assert isinstance(resp, bool)

    def test_get_pastes_raise_if_email_not_specified(self):
        # pyHIBP.get_pastes():
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.get_pastes()
        assert "The email address supplied must be provided, and be a string" in str(excinfo.value)

    def test_get_pastes_raise_if_email_not_string(self):
        # pyHIBP.get_pastes(email_address=1):
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.get_pastes(email_address=1)
        assert "The email address supplied must be provided, and be a string" in str(excinfo.value)


class TestGetDataClasses(object):
    def test_get_data_classes(self):
        # get_data_classes():
        resp = pyHIBP.get_data_classes()
        assert isinstance(resp, list)
        assert len(resp) > 10
        assert "Passwords" in resp


class TestIsPasswordBreached(object):
    """ NB: Deprecated function ... moved to pwnedpasswords.is_password_breached """

    def test_is_password_breached_password_only_breached(self):
        # is_password_breached(password=TEST_PASSWORD, sha1_hash=None):
        assert pyHIBP.is_password_breached(password=TEST_PASSWORD)

    def test_is_password_breached_sha1hash_only_breached(self):
        # is_password_breached(password=None, sha1_hash=TEST_PASSWORD_SHA1_HASH):
        assert pyHIBP.is_password_breached(sha1_hash=TEST_PASSWORD_SHA1_HASH)

    def test_is_password_breached_password_only_not_breached(self):
        # is_password_breached(password=TEST_PASSWORD_LIKELY_NOT_COMPROMISED, sha1_hash=None):
        assert not pyHIBP.is_password_breached(password=TEST_PASSWORD_LIKELY_NOT_COMPROMISED)

    def test_is_password_breached_sha1hash_only_not_breached(self):
        # is_password_breached(password=None, sha1_hash=TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH):
        assert not pyHIBP.is_password_breached(sha1_hash=TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH)

    def test_is_password_breached_password_and_sha1hash_matches(self):
        # is_password_breached(password=TEST_PASSWORD, sha1_hash=TEST_PASSWORD_SHA1_HASH):
        assert pyHIBP.is_password_breached(password=TEST_PASSWORD, sha1_hash=TEST_PASSWORD_SHA1_HASH)

    def test_is_password_breached_raise_if_no_params_specified(self):
        # is_password_breached(password=None, sha1_hash=None)
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.is_password_breached()
        assert "You must provide either a password or sha1_hash" in str(excinfo.value)

    def test_is_password_breached_raise_if_password_not_string(self):
        # is_password_breached(password=1, sha1_hash=None)
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.is_password_breached(password=1)
        assert "The provided password is not a string" in str(excinfo.value)

    def test_is_password_breached_raise_if_sha1hash_not_string(self):
        # is_password_breached(password=None, sha1_hash=1)
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.is_password_breached(sha1_hash=1)
        assert "The provided sha1_hash is not a string" in str(excinfo.value)

    def test_is_password_breached_raise_if_password_and_sha1hash_mismatch(self):
        # is_password_breached(password="NotThePassword", sha1_hash=TEST_PASSWORD_SHA1_HASH):
        with pytest.raises(AttributeError) as excinfo:
            pyHIBP.is_password_breached(password="NotThePassword", sha1_hash=TEST_PASSWORD_SHA1_HASH)
        assert "A password and SHA1 hash were supplied (only one is needed), but they did not match" in str(excinfo.value)


class TestMiscellaneous(object):
    def test_raise_if_rate_limit_exceeded(self):
        """
        The API will respond the same to all exceeded rate limits across all endpoints; only need to test this
        once.
        """
        # (x2) get_account_breaches(account=None, domain=None, truncate_response=True, include_unverified=False):
        with pytest.raises(RuntimeError) as excinfo:
            pyHIBP.get_account_breaches(account=TEST_ACCOUNT, truncate_response=True)
            pyHIBP.get_account_breaches(account=TEST_ACCOUNT, truncate_response=True)
        assert "HTTP 429" in str(excinfo.value)

    def test_raise_if_useragent_is_not_set(self, monkeypatch):
        # This should never be encountered normally, since we have the module-level variable/constant;
        # That said, test it, since we can, and since we might as well cover the line of code.
        monkeypatch.setattr(pyHIBP, 'pyHIBP_USERAGENT', None)
        with pytest.raises(RuntimeError) as excinfo:
            pyHIBP.get_data_classes()
        assert "HTTP 403" in str(excinfo.value)

    def test_raise_if_invalid_format_submitted(self):
        # For example, if a null (0x00) character is submitted to an endpoint.
        with pytest.raises(RuntimeError) as execinfo:
            pyHIBP.get_account_breaches(account="\0")
        assert "HTTP 400" in str(execinfo.value)


test/test_pwnedpasswords.py
import hashlib
import time

import pytest

from pyHIBP import pwnedpasswords as pw


TEST_PASSWORD = "password"
TEST_PASSWORD_SHA1_HASH = hashlib.sha1(TEST_PASSWORD.encode('utf-8')).hexdigest()
# At least, I doubt someone would have used this (only directly specifying here for deterministic tests...)
TEST_PASSWORD_LIKELY_NOT_COMPROMISED = "&Q?t{%i|n+&qpyP/`/Lyr3<rK|N/M//;u^!fnR+j'lM)A+IGcgRGs[6mLY7yV-|x0bYB&L.JyaJ"
TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH = hashlib.sha1(TEST_PASSWORD_LIKELY_NOT_COMPROMISED.encode('utf-8')).hexdigest()


@pytest.fixture(autouse=True)
def rate_limit():
    # There's no rate limit on passwords, but be nice anyway.
    time.sleep(1)


class TestIsPasswordBreached(object):
    def test_is_password_breached_password_only_breached(self):
        # is_password_breached(password=TEST_PASSWORD, sha1_hash=None):
        resp = pw.is_password_breached(password=TEST_PASSWORD)
        assert isinstance(resp, int)
        # It's actually 3303003 at last check, but 100 works for PoC.
        assert resp > 100

    def test_is_password_breached_sha1hash_only_breached(self):
        # is_password_breached(password=None, sha1_hash=TEST_PASSWORD_SHA1_HASH):
        resp = pw.is_password_breached(sha1_hash=TEST_PASSWORD_SHA1_HASH)
        assert isinstance(resp, int)

    def test_is_password_breached_password_only_not_breached(self):
        # is_password_breached(password=TEST_PASSWORD_LIKELY_NOT_COMPROMISED, sha1_hash=None):
        resp = pw.is_password_breached(password=TEST_PASSWORD_LIKELY_NOT_COMPROMISED)
        assert isinstance(resp, int)
        assert resp == 0

    def test_is_password_breached_sha1hash_only_not_breached(self):
        # is_password_breached(password=None, sha1_hash=TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH):
        resp = pw.is_password_breached(sha1_hash=TEST_PASSWORD_LIKELY_NOT_COMPROMISED_HASH)
        assert isinstance(resp, int)
        assert resp == 0

    def test_is_password_breached_password_and_sha1hash_matches(self):
        # is_password_breached(password=TEST_PASSWORD, sha1_hash=TEST_PASSWORD_SHA1_HASH):
        resp = pw.is_password_breached(password=TEST_PASSWORD, sha1_hash=TEST_PASSWORD_SHA1_HASH)
        assert isinstance(resp, int)
        assert resp > 100

    def test_is_password_breached_raise_if_no_params_specified(self):
        # is_password_breached(password=None, sha1_hash=None)
        with pytest.raises(AttributeError) as excinfo:
            pw.is_password_breached()
        assert "You must provide either a password or sha1_hash" in str(excinfo.value)

    def test_is_password_breached_raise_if_password_not_string(self):
        # is_password_breached(password=1, sha1_hash=None)
        with pytest.raises(AttributeError) as excinfo:
            pw.is_password_breached(password=1)
        assert "The provided password is not a string" in str(excinfo.value)

    def test_is_password_breached_raise_if_sha1hash_not_string(self):
        # is_password_breached(password=None, sha1_hash=1)
        with pytest.raises(AttributeError) as excinfo:
            pw.is_password_breached(sha1_hash=1)
        assert "The provided sha1_hash is not a string" in str(excinfo.value)

    def test_is_password_breached_raise_if_password_and_sha1hash_mismatch(self):
        # is_password_breached(password="NotThePassword", sha1_hash=TEST_PASSWORD_SHA1_HASH):
        with pytest.raises(AttributeError) as excinfo:
            pw.is_password_breached(password="NotThePassword", sha1_hash=TEST_PASSWORD_SHA1_HASH)
        assert "A password and SHA1 hash were supplied (only one is needed), but they did not match" in str(excinfo.value)


class TestRangeSearch(object):
    def test_no_params_provided_raises(self):
        # range_search(password=None, first_5_hash_chars=None, sha1_hash=None):
        with pytest.raises(AttributeError) as execinfo:
            pw.range_search()
        assert "One of password, first_5_hash_chars, or sha1_hash must be provided." in str(execinfo.value)

    def test_password_not_string_raises(self):
        # range_search(password=123, first_5_hash_chars=None, sha1_hash=None):
        with pytest.raises(AttributeError) as execinfo:
            pw.range_search(password=123)
        assert "password must be a string type." in str(execinfo.value)

    def test_first_5_hash_chars_not_string_raises(self):
        # range_search(password=None, first_5_hash_chars=123, sha1_hash=None):
        with pytest.raises(AttributeError) as execinfo:
            pw.range_search(first_5_hash_chars=123)
        assert "first_5_hash_chars must be a string type." in str(execinfo.value)

    def test_first_5_hash_chars_not_length_five_raises(self):
        # range_search(password=None, first_5_hash_chars="123456", sha1_hash=None):
        with pytest.raises(AttributeError) as execinfo:
            pw.range_search(first_5_hash_chars="123456")
        assert "first_5_hash_chars must be of length 5." in str(execinfo.value)

    def test_sha1_hash_not_string_raises(self):
        # range_search(password=None, first_5_hash_chars=None, sha1_hash=123):
        with pytest.raises(AttributeError) as execinfo:
            pw.range_search(sha1_hash=123)
        assert "sha1_hash must be a string type." in str(execinfo.value)

    def test_list_of_partial_hashes_returned_with_5chars(self):
        # range_search(password=None, first_5_hash_chars=TEST_PASSWORD_SHA1_HASH[0:5], sha1_hash=None):
        resp = pw.range_search(first_5_hash_chars=TEST_PASSWORD_SHA1_HASH[0:5])
        assert isinstance(resp, list)
        assert len(resp) > 100
        match_found = False
        for entry in resp:
            if TEST_PASSWORD_SHA1_HASH[5:] in entry.lower():
                match_found = True
                break
        assert match_found

    def test_provide_password_to_function(self):
        resp = pw.range_search(password="password")
        assert isinstance(resp, int)
        assert resp > 100

    def test_ensure_case_sensitivity_of_hash_does_not_matter(self):
        resp_one = pw.range_search(sha1_hash=TEST_PASSWORD_SHA1_HASH.lower())
        assert isinstance(resp_one, int)
        assert resp_one > 100

        resp_two = pw.range_search(sha1_hash=TEST_PASSWORD_SHA1_HASH.upper())
        assert isinstance(resp_two, int)
        assert resp_two > 100

        assert resp_one == resp_two

    def test_zero_count_result_for_non_breached_password(self):
        resp = pw.range_search(password=TEST_PASSWORD_LIKELY_NOT_COMPROMISED)
        assert isinstance(resp, int)
        assert resp == 0


src/pyHIBP/__init__.py
import warnings

import requests
import six

from pyHIBP import pwnedpasswords as pw

HIBP_API_BASE_URI = "https://haveibeenpwned.com/api/v2/"
HIBP_API_ENDPOINT_BREACH_SINGLE = "breach/"
HIBP_API_ENDPOINT_BREACHES = "breaches"
HIBP_API_ENDPOINT_BREACHED_ACCT = "breachedaccount/"
HIBP_API_ENDPOINT_DATA_CLASSES = "dataclasses"
HIBP_API_ENDPOINT_PASTES = "pasteaccount/"

# The HIBP API requires that a useragent be set.
pyHIBP_USERAGENT = "pyHIBP: A Python Interface to the Public HIBP API"


def _process_response(response):
    """
    Process the `requests` response from the call to the HIBP API endpoints.

    :param response: The response object from a call to `requests`
    :return: True if HTTP Status 200, False if 404. Raises RuntimeError on API-defined status codes of
    400, 403, 429; NotImplementedError if the API returns an unexpected HTTP status code.
    """
    if response.status_code == 200:
        # The request was successful (a password/breach/paste was found)
        return True
    elif response.status_code == 404:
        # The request was successful, though the item wasn't found
        return False
    elif response.status_code == 400:
        # Bad request - The account does not comply with an acceptable format (i.e., it's an empty string)
        raise RuntimeError(
            "HTTP 400 - Bad request - The account does not comply with an acceptable format (i.e., it's an empty string)")
    elif response.status_code == 403:
        # Forbidden - no user agent has been specified in the request
        raise RuntimeError("HTTP 403 - User agent required for HIBP API requests, but no user agent was sent to the API endpoint")
    elif response.status_code == 429:
        # Too many requests - the rate limit has been exceeded
        raise RuntimeError(
            "HTTP 429 - Rate limit exceeded: API rate limit is 1500ms. Retry-After header was: " + response.headers['Retry-After']
        )
    else:
        # We /should/ get one of the above error codes. If not, raise an error.
        raise NotImplementedError("Returned HTTP status code of " + str(response.status_code) + " was not expected.")


def get_account_breaches(account=None, domain=None, truncate_response=False, include_unverified=False):
    """
    Gets breaches for a specified account from the HIBP system, optionally restricting the returned results
    to a specified domain.

    :param account: The user's account name (such as an email address or a user-name). Default None.
    :param domain: The domain to check for breaches. Default None.
    :param truncate_response: If ``account`` is specified, truncates the response down to the breach names.
    Default False.
    :param include_unverified: If set to True, unverified breaches are included in the result. Default False.
    :return: A list object containing one or more dict objects, based on the information being requested,
    provided there was matching information. Boolean False returned if no information was found according to
    the HIBP API.
    """
    # Account/Domain don't need to be specified, but they must be text if so.
    if account is None or not isinstance(account, six.string_types):
        raise AttributeError("The account parameter must be specified, and must be a string")
    if domain is not None and not isinstance(domain, six.string_types):
        raise AttributeError("The domain parameter, if specified, must be a string")

    # Build the URI
    uri = HIBP_API_BASE_URI + HIBP_API_ENDPOINT_BREACHED_ACCT + account
    headers = {'user-agent': pyHIBP_USERAGENT}

    # Build the query string payload (requests drops params when None)
    # (and the HIBP backend ignores those that don't apply)
    query_string_payload = {
        "domain": domain,
        "truncateResponse": truncate_response,
        "includeUnverified": include_unverified,
    }
    resp = requests.get(url=uri, params=query_string_payload, headers=headers)
    if _process_response(response=resp):
        return resp.json()
    else:
        return False


def get_all_breaches(domain=None):
    """
    Returns a listing of all sites breached in the HIBP database.

    :param domain: Optional, default None. If specified, get all breaches for the domain with the specified name.
    :return: A list object containing one or more dict objects if breaches are present. Returns Boolean False
    if ``domain`` is specified, but the resultant list would be length zero.
    """
    if domain is not None and not isinstance(domain, six.string_types):
        raise AttributeError("The domain parameter, if specified, must be a string")
    uri = HIBP_API_BASE_URI + HIBP_API_ENDPOINT_BREACHES
    headers = {'user-agent': pyHIBP_USERAGENT}
    query_string_payload = {'domain': domain}
    resp = requests.get(url=uri, params=query_string_payload, headers=headers)
    # The API will return HTTP200 even if resp.json is length zero.
    if _process_response(response=resp) and len(resp.json()) > 0:
        return resp.json()
    else:
        return False


def get_single_breach(breach_name=None):
    """
    Returns a single breach's information from the HIBP's database.

    :param breach_name: The breach to retrieve. Required.
    :return: A dict object containing the information for the specified breach name, if it exists in the HIBP
    database. Boolean False is returned if the specified breach was not found.
    """
    if not isinstance(breach_name, six.string_types):
        raise AttributeError("The breach_name must be specified, and be a string")
    uri = HIBP_API_BASE_URI + HIBP_API_ENDPOINT_BREACH_SINGLE + breach_name
    headers = {'user-agent': pyHIBP_USERAGENT}
    resp = requests.get(url=uri, headers=headers)
    if _process_response(response=resp):
        return resp.json()
    else:
        return False


def get_pastes(email_address=None):
    """
    Retrieve all pastes for a specified email address.

    :param email_address: The email address to search. Required.
    :return: A list object containing one or more dict objects corresponding to the pastes the specified email
    address was found in. Boolean False returned if no pastes are detected for the given account.
    """
    if not isinstance(email_address, six.string_types):
        raise AttributeError("The email address supplied must be provided, and be a string")
    uri = HIBP_API_BASE_URI + HIBP_API_ENDPOINT_PASTES + email_address
    headers = {'user-agent': pyHIBP_USERAGENT}
    resp = requests.get(url=uri, headers=headers)
    if _process_response(response=resp):
        return resp.json()
    else:
        return False


def get_data_classes():
    """
    Retrieves all available data classes from the HIBP API.

    :return: A list object containing available data classes, corresponding to attributes found in breaches.
    A given breach will have one or more of the data classes in the list.
    """
    uri = HIBP_API_BASE_URI + HIBP_API_ENDPOINT_DATA_CLASSES
    headers = {'user-agent': pyHIBP_USERAGENT}
    resp = requests.get(url=uri, headers=headers)
    if _process_response(response=resp):
        return resp.json()
    else:
        # This path really shouldn't return false
        raise RuntimeError("HIBP API returned HTTP404 on a request for data classes.")


def is_password_breached(password=None, sha1_hash=None):
    """
    __DEPRECATED__: Use ``pwnedpasswords.is_password_breached`` instead, which contains additional functionality.

    Checks the HIBP breached password corpus for a breached password. Only the password or sha1_hash
    parameter is required to be set.

    Note that while the HIBP endpoint does have a originalPasswordIsAHash parameter, passwords submitted
    to this function will successfully process a supplied SHA1 hash password, since we pre-hash on our end.

    :param password: The raw password to check. Will be converted to a SHA1 hash prior to submission.
    :param sha1_hash: The SHA1 hash of the password to check.
    :return: True if the password was in the HIBP password corpus, otherwise False.
    """
    warnings.warn("Deprecation Warning: is_password_breached has moved to the pwnedpasswords module.")

    # Partially re-implement parameter checks to handle this deprecated function wrapper.
    if not password and not sha1_hash:
        raise AttributeError("You must provide either a password or sha1_hash.")

    # Pass the variables through to the new function...
    resp = pw.is_password_breached(password=password, sha1_hash=sha1_hash)

    if resp:
        # If the response is greater than zero
        return True
    else:
        return False
